src/backend/controllers/crowdsourcer/crowdsourcer.ts

```diff
@@ -1,9 +1,39 @@
 import { Response } from 'express';
 import { omit } from 'lodash';
 import { crowdsourcerSchema } from '../../models/Crowdsourcer';
+import { referralSchema } from '../../models/Referral';
 import { handleQueries } from '../utils';
 import * as Interfaces from '../utils/interfaces';
 
+export const createReferral = async (req: Interfaces.EditorRequest, res: Response): Promise<void> => {
+  const {
+    mongooseConnection,
+    referralCode,
+    user: { uid },
+  } = handleQueries(req);
+
+  const Crowdsourcer = mongooseConnection.model<Interfaces.Crowdsourcer>('Crowdsourcer', crowdsourcerSchema);
+
+  const [referrer, referredUser] = await Promise.all([
+    Crowdsourcer.findOne({ referralCode }),
+    Crowdsourcer.findOne({ firebaseId: uid }),
+  ]);
+
+  const Referral = mongooseConnection.model<Interfaces.Referral>('Referral', referralSchema);
+
+  const existingReferral = await Referral.findOne({ referredUserId: referredUser.id });
```

Comment:
instead of using the document Ids, we want to use the Firebase uid to make it easier to look up specific users
```suggestion
  const existingReferral = await Referral.findOne({ referredUserId: referredUser.firebaseId });
```

Reply:
I would push back on this here. We don't want to tightly couple our internal `Crowdsourcer` usage with firebase as it wont provide portability and other issues in the future. Firebase id is merely a bridge between firebase auth and `Crowdsourcer` hence it should serve no other purpose. Besides it was described here in the [tech documentation](https://docs.google.com/document/d/1rP9LiRbWmVWxlGsr3PKxcP8_PL9Y7Siuy1MUfkGrLHM/edit#heading=h.6mj7fbesgj10) prior to implementation.

Reply:
Specific users can be looked up using a [$lookup aggregation](https://www.mongodb.com/docs/manual/reference/operator/aggregation/lookup/) query

Reply:
Thanks for mentioning the design document! I definitely missed that point. However, using the Firebase Id has been a consistent pattern across all our projects and has shown to scale very well since we depend on the unique user ids that are already generated to prevent duplicating the concept of having a 1:1 relationship between a distinct user and their identification code. If we use the Crowdsourcer Id to associate with a user, then we are duplicating this relationship which can lead to more code complexity or maintainance. Could you explain more in detail what you mean by "it won't provide portability and other issues in the future" specifically?

Additionally, `$lookup` aggregation has performance concerns that only really arise when we have the alternative of being able to create simplistic queries that use fields that are indexed without the extra lookup step.

Reply:
To elaborate further on`'it won't provide portability and other may cause other issues in the future'` i meant if we decide to pick a different authentication provider in the future (given how often software changes) it would be difficult to update the changes across the entire project. It may even involve a db migration. However decoupling `Crowdsourcer` id from firebase id will provide flexibility in the future. If firebase id is already implemented project wide then i guess it only makes sense to be consistent.

Gotcha!